### PR TITLE
Remove invalid implicit downcast

### PR DESCRIPTION
**Pull request by sharwell**
Currently `SeparatedSyntaxList<SyntaxNode>` can be implicitly cast to `SeparatedSyntaxList<T>` for *any syntax type `T`*. This is not a safe or meaningful conversion, so this change seeks to remove it.

* The implicit operator is left in place using its underlying metadata signature to preserve *binary* compatibility with previous builds
* The implicit operator is marked obsolete to prevent future use of this unsafe method
* ~~No explicit operator could be added because the implicit operator was not removed, so a `CastDown` method is added to provide the functionality via an explicit method~~ An explicit operator alongside the `op_Implicit` method using the operator syntax.

### CODE REVIEW

File: src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs [hunk 1]
```diff
@@ -592,9 +592,16 @@ IEnumerator IEnumerable.GetEnumerator()
...
+        [Obsolete("Implicit downcast is not safe. Use CastDown<TNode>() instead", error: true)]
         public static implicit operator SeparatedSyntaxList<TNode>(SeparatedSyntaxList<SyntaxNode> nodes)
         {
             return new SeparatedSyntaxList<TNode>(nodes._list);
         }
+
+        public SeparatedSyntaxList<TDerived> CastDown<TDerived>()
+            where TDerived : TNode
+        {
+            return new SeparatedSyntaxList<TDerived>(_list);
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs]: Note: this is still not safe.  Which is why we didn't go this way originally.

Comment (sharwell) [inline src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs]: @CyrusNajmabadi However, the *API* provides a safe alternative. Since this type is immutable, we have the option of casting each element of `_list` here to force an `InvalidCastException` as early as possible, but I have a hard time balancing that against the performance of things like `ISyntaxFactsService` which always return an upcast list.

Would renaming this method to `UncheckedCastDown` resolve your concern?

Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs]: > Since this type is immutable, we have the option of casting each element of _list here to force an InvalidCastException as early as possible

Right, we discarded that as an option due to not wanting a cast to be O(n) when n could be quite large.  

> Would renaming this method to UncheckedCastDown resolve your concern?

Possibly.  We'd need to take this through API review (for the new api), and compat council (for the obsoleting).  I wouldn't be opposed.  But i'm not strongly in favor either.

--

Also note: we'd need to do the same for SyntaxList as well, right?

**Comment by jcouv** [comment]
Given that there was no activity on this PR in a while, I'm marking it as "for personal review" for now (so it doesn't show up in our daily list of PRs to review). Thanks

**Comment by sharwell** [comment]
@jcouv this has been updated again

File: src/CodeStyle/Core/Analyzers/Extensions/CodeStyleSeparatedSyntaxListExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+            return CastDownHelper<TDerived>(list);
+#pragma warning restore CS0612 // Type or member is obsolete
+        }
+
+        [Obsolete]
+        private static SeparatedSyntaxList<TDerived> CastDownHelper<TDerived>(SeparatedSyntaxList<SyntaxNode> list)
+            where TDerived : SyntaxNode
+        {
+            return list;
+        }
```
Comment (CyrusNajmabadi) [inline src/CodeStyle/Core/Analyzers/Extensions/CodeStyleSeparatedSyntaxListExtensions.cs:24]: slightly confused by this pattern.  can this all be replaced with:

```c#
        public static SeparatedSyntaxList<TDerived> CastDown<TDerived>(this SeparatedSyntaxList<SyntaxNode> list)
            where TDerived : SyntaxNode
        {
#pragma warning disable CS0612 // Type or member is obsolete
            return (SeparatedSyntaxList<TDerived>)list;
#pragma warning restore CS0612 // Type or member is obsolete
        }
```

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/CodeStyle/Core/Analyzers/Extensions/CodeStyleSeparatedSyntaxListExtensions.cs (hunk 2 above)
Comment (sharwell) [reply src/CodeStyle/Core/Analyzers/Extensions/CodeStyleSeparatedSyntaxListExtensions.cs:24]: I tried that, but it didn't work. I don't think it's possible to use `#pragma warning disable` with `error: true`.

Comment (CyrusNajmabadi) [reply src/CodeStyle/Core/Analyzers/Extensions/CodeStyleSeparatedSyntaxListExtensions.cs:24]: oh interesting.  the pattern makes sense now then :)

**Comment by AlekseyTs** [comment]
>* The implicit operator is left in place to preserve *binary* compatibility with previous builds
>* The implicit operator is marked obsolete to prevent future use of this unsafe method
>* No explicit operator could be added because the implicit operator was not removed, so a `CastDown` method is added to provide the functionality via an explicit method

Have you considered replacing the implicit operator with a regular method with the same name an signature instead? That should take care of binary compatibility. Then we can add an explicit operator.

**Comment by sharwell** [comment]
> Have you considered replacing the implicit operator with a regular method with the same name an signature instead? That should take care of binary compatibility. Then we can add an explicit operator.

💭 No, but I'll try now

**Comment by CyrusNajmabadi** [comment]
> Have you considered replacing the implicit operator with a regular method with the same name an signature instead? That should take care of binary compatibility. Then we can add an explicit operator.

TIL! :-)

**Comment by sharwell** [comment]
@AlekseyTs Now updated to preserve the implicit operator using its metadata name instead of `implicit operator`.

File: src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs [hunk 3]
```diff
@@ -598,7 +599,14 @@ IEnumerator IEnumerable.GetEnumerator()
             return new SeparatedSyntaxList<SyntaxNode>(nodes._list);
         }
 
-        public static implicit operator SeparatedSyntaxList<TNode>(SeparatedSyntaxList<SyntaxNode> nodes)
+        [Obsolete("Implicit downcast is not safe. Use explicit cast instead.", error: true)]
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs:602]: Since this is no longer an operator. Consider to not refer to it as one. Something like: "This method is not safe and present for binary compatibility only. Use explicit cast instead."

File: src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs [hunk 4]
```diff
@@ -448,7 +449,9 @@ public override int GetHashCode()
             return _node?.GetHashCode() ?? 0;
         }
 
-        public static implicit operator SyntaxList<TNode>(SyntaxList<SyntaxNode> nodes)
+        [Obsolete("Implicit downcast is not safe. Use explicit cast instead.", error: true)]
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Syntax/SyntaxList`1.cs:452]: Similar suggestion

File: src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxRewriter.cs [hunk 5]
```diff
@@ -138,24 +138,24 @@ public virtual SyntaxTrivia VisitTrivia(SyntaxTrivia trivia)
 
         public virtual SyntaxList<TNode> VisitList<TNode>(SyntaxList<TNode> list) where TNode : SyntaxNode
         {
-            SyntaxListBuilder? alternate = null;
+            SyntaxListBuilder<TNode> alternate = default;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxRewriter.cs:141]: Is this change primarily related to the stated goal of the PR? If not, please consider separating it into its own PR.

**Review by AlekseyTs — APPROVED** [review]
Changes under Compilers LGTM, modulo possibly unrelated change in CSharpSyntaxRewriter.cs

File: src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs (hunk 3 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs]: Now updated.

File: src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/Syntax/SeparatedSyntaxList.cs]: This has been resolved by simply changing the implicit cast to an explicit cast.

**Review by 333fred — COMMENTED** [review]
Changes under compilers look fine, but not signing off until build and test issues are corrected.

**Comment by jaredpar** [comment]
@sharwell this PR has appropriate sign offs. It's a bit stale so we probably want to re-run CI here. But are you intending to move forward with this? If not I would like to close to get it off our review queue.

**Comment by jaredpar** [comment]
Have not heard back so moving to draft.

**Comment by sharwell** [comment]
@jaredpar Sorry got hit by the command line build failures when I tried to update. Trying again locally and hopefully should work out.
